app/api/ada/adaTransactions/adaNewTransactions.js

```diff
@@ -46,107 +28,149 @@ const protocolMagic = CONFIG.network.protocolMagic;
+
   const feeAlgorithm = RustModule.Wallet.LinearFeeAlgorithm.default();
+  // firts build a transaction to see what the cost would be
+  const fakeTxBuilder = new RustModule.Wallet.TransactionBuilder();
+  const inputs = utxoToTxInput(allUtxos);
+  addTxInputs(fakeTxBuilder, inputs);
+  const inputAmount = coinToBigNumber(fakeTxBuilder.get_input_total());
+  addOutput(fakeTxBuilder, receiver, inputAmount.toString());
+  const fee = coinToBigNumber(fakeTxBuilder.estimate_fee(feeAlgorithm));
 
-  const allAdaAddresses = await getAdaAddressesList();
-  const allUtxos = await getAllUTXOsForAddresses(allAdaAddresses.map(addr => addr.cadId));
-  const txInputs = utxoToTxInput(allUtxos);
-  let selectionResult;
-  try {
-    selectionResult = getInputSelection(txInputs, outputPolicy, feeAlgorithm, receiver, amount);
-  } catch (err) {
+  // create a new transaction subtracing the fee from your total UTXO
+  const newAmount = totalBalance.minus(fee);
+  const unsignedTxResponse = await newAdaUnsignedTxFromUtxo(receiver, newAmount, null, allUtxos);
+
+  // sanity check
+  const balance = unsignedTxResponse.txBuilder.get_balance(feeAlgorithm);
+  if (balance.is_negative()) {
+    /**
+     * The balance may be slightly positive. This is because lowering the "amount" to send
+     * May reduce the amount of bytes required for the "amount", causing the fee to also drop
+     */
     throw new NotEnoughMoneyToSendError();
   }
-  const senderInputs = txInputs.filter(input => (
-    selectionResult.is_input(
-      RustModule.Wallet.TxoPointer.from_json(input.to_json().ptr)
-    )
-  ));
 
-  const usedAddressSet = new Set(senderInputs.map(input => input.to_json().value.address));
-  const usedAdaAddresses = allAdaAddresses.filter(
+  return unsignedTxResponse;
+}
+
+/**
+ * @param {*} possibleInputAddresses we send all UTXO associated with an address.
+ * This maximizes privacy.
+ */
+export async function newAdaUnsignedTx(
+  receiver: string,
+  amount: string,
+  changeAdaAddr: ?AdaAddress,
+  possibleInputAddresses: Array<AdaAddress>,
+  addressesToUtxos: Array<string> => Promise<Array<UTXO>>
+): Promise<UnsignedTxResponse> {
+  const allUtxos = await addressesToUtxos(possibleInputAddresses.map(addr => addr.cadId));
+  const unsignedTxResponse = await newAdaUnsignedTxFromUtxo(
+    receiver,
+    amount,
+    changeAdaAddr,
+    allUtxos
+  );
+
+  const usedAddressSet = new Set(unsignedTxResponse.senderUtxos.map(utxo => utxo.receiver));
+  const usedAdaAddresses = possibleInputAddresses.filter(
     adaAddress => usedAddressSet.has(adaAddress.cadId)
   );
 
+  const addressesMap = addressesToAddressMap(usedAdaAddresses);
+
+  return {
+    ...unsignedTxResponse,
+    addressesMap
+  };
+}
+
+export async function newAdaUnsignedTxFromUtxo(
+  receiver: string,
+  amount: string,
+  changeAdaAddr: ?AdaAddress,
+  allUtxos: Array<UTXO>,
+): Promise<UnsignedTxFromUtxoResponse> {
+  const feeAlgorithm = RustModule.Wallet.LinearFeeAlgorithm.default();
```

Comment:
it’s a really small detail but I would move this right before it’s used. We may even not actually use it and we still invoke it.

Reply:
nvm